### PR TITLE
Wipe before reset/recovery and introduce 'initialized' field

### LINKED ISSUE #1065: Wipe before reset and recovery
As a follow-up of #1063 and #1056 we have agreed with @matejcik it would be nice to wipe the storage before reset/recovery to make sure the storage is in an expected state. Although the device should not receive any settings after #1063 this seems as a good precaution.

Not to confuse the Host we want to keep the device id unchanged.

1. We will introduce a new function which will be called before reset/recovery:

```
reset_storage():  # name to be determined
- remember device_id if any
- wipe
- storage init
- write the previous device id
- write storage version
```

2. We will add a new public field `initialized` which will be used in Features.

### PR DESCRIPTION
**Pull request by tsusanka**
I could use some more eyes on this to see what you think (about the migration most importantly). So I am filing a draft PR to make sure this is not merged before reviewed.

----

This fixes #1065. Please read the issue for rationale and more details.

This introduces a new public storage field "initialized" denoting whether the device has mnemonic or not. We have used "storage version" for that previously but that was a private field. We could make the field public but this seems as a clearer way.

An important question is how to migrate the old storage. I have finally bumped the storage version but I want to know what you think.

- All older firmwares migrate from the 02 version using `_migrate_from_version_02`. Since the version was previously written [after](https://github.com/trezor/trezor-firmware/blob/7364883c04fd5cafb4b40407c2f2d78521624112/core/src/storage/device.py#L122) the device was restored we know the device is initialized when migrating.
- For the uninitialized devices the migration does not happen so that seems correct as well. The field is not present so `initialized=False`. The version gets then written during `storage.reset` and the `initialized` field in `store_mnemonic_secret`.
- Due to the fact that the version is now written directly after wipe migrating in the opposite direction (to 2.3.0 for example) with uninitialized device will incorrectly claim it is initialized. We can still set the storage version after initialization to fix this but it seems weird. Why would anyone need to downgrade with an uninitialized device anyway?

### CODE REVIEW

**Comment by andrewkozlik** [comment]
> I suggest we convert `get_version` to work similarly to how `get_device_id` works: if the version field is unset, it should be set (and written) to current version.

I guess this makes the most sense of the possibilities at hand.

> Could we repurpose field `0x01` to mean `INITIALIZED` (and convert it to public), and move `_VERSION` to `0x13`?

I don't see what advantage this brings.

**Comment by matejcik** [comment]
> > Could we repurpose field `0x01` to mean `INITIALIZED` (and convert it to public), and move `_VERSION` to `0x13`?
> 
> I don't see what advantage this brings.

It would allow for full backwards compatibility: older firmwares would see `VERSION` set iff the device is initialized.

Reasons why this *probably* doesn't work:
* `TRUE_BYTE` is `0x01`, which would get read as version 1 of storage, so the U2F counter would get upgraded (wrongly)
* `INITIALIZED` is supposed to be public, `_VERSION` isn't. ISTM public and non-public fields cannot be cross-read, so trying to read field 0x01 as private would fail.

**Comment by andrewkozlik** [comment]
> Reasons why this _probably_ doesn't work:
> 
> * `TRUE_BYTE` is `0x01`, which would get read as version 1 of storage, so the U2F counter would get upgraded (wrongly)
> 
> * `INITIALIZED` is supposed to be public, `_VERSION` isn't. ISTM public and non-public fields cannot be cross-read, so trying to read field 0x01 as private would fail.

Correct. It breaks downgrade even more, we would have to forbid firmware downgrade altogether.

**Comment by tsusanka** [comment]
> Currently the storage version is only set on a fully initialized device. In context of Shamir persistent recovery, that seems incorrect.

I do not think the storage version is stored when doing Shamir persistent recovery. Are you saying it is?

File: core/src/storage/device.py [hunk 1]
```diff
@@ -60,6 +61,10 @@ def set_version(version: bytes) -> None:
     common.set(_NAMESPACE, _VERSION, version)
 
 
+def is_initialized() -> bool:
```
[the remark this answers is not in the document]
Comment (tsusanka) [reply core/src/storage/device.py:64]: Well, it just seemed to fit the idea of distinguishing between the storage being initialized and the fact that the device is.

**Comment by tsusanka** [comment]
I tried the downgrade and on 09e7b05 the downgrade actually works just fine. That is due to the fact that we write the current version either after `reset()` or after the migration but not during `init_unlocked`. This is different when e7fca798a8f4241c7c2e41722796e532f99b2158 is added. In that case we write the version when `get_version` is called which means also for `init_unlocked`.

---

Anyway, it seems we currently have number of solutions.

1. e7fca798a8f4241c7c2e41722796e532f99b2158 and forbid downgrading (in another words enforce wipe before downgrade).
2. 09e7b05 which brings inconsistency because the storage version is written after upgrade (migration) and after `reset` but not after `wipe`. This is incorrect.
3. Write the version as we used that means after reset/recovery is finished.

Although it seems (1) is the cleanest solution I believe (3) is just fine and we are sure we do not break anything.

**Comment by tsusanka** [comment]
I have opened new PR to simplify the review: #1072.

**Pull request closed without merging**